File: app/src/main/java/net/ducksmanager/whattheduck/ItemList.java
package net.ducksmanager.whattheduck;


import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import com.github.clans.fab.FloatingActionButton;
import com.github.clans.fab.FloatingActionMenu;

import net.ducksmanager.util.CoverFlowFileHandler;
import net.ducksmanager.util.Settings;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import static android.view.View.GONE;
import static android.view.View.INVISIBLE;
import static android.view.View.VISIBLE;

public abstract class ItemList<Item> extends AppCompatActivity {
    public static String type = WhatTheDuck.CollectionType.USER.toString();
    static final int MIN_ITEM_NUMBER_FOR_FILTER = 20;
    private static final int REQUEST_IMAGE_CAPTURE = 1;

    private Boolean requiresDataDownload = false;
    protected List<Item> data = new ArrayList<>();

    protected abstract boolean hasList();
    protected abstract void downloadList();
    protected abstract boolean hasDividers();

    protected abstract boolean isPossessedByUser();
    protected abstract void setData();

    protected abstract boolean shouldShow();
    protected abstract boolean shouldShowNavigationCountry();
    protected abstract boolean shouldShowNavigationPublication();
    protected abstract boolean shouldShowToolbar();
    protected abstract boolean shouldShowAddToCollectionButton();
    protected abstract boolean shouldShowFilter(List<Item> items);

    protected abstract ItemAdapter<Item> getItemAdapter();

    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.wtd_list);

        View navigationAllCountries = this.findViewById(R.id.navigationAllCountries);
        if (navigationAllCountries != null) {
            navigationAllCountries.setOnClickListener(view ->
                goToView(CountryList.class)
            );
        }

        View navigationCurrentCountry = this.findViewById(R.id.navigationCountry);
        if (navigationCurrentCountry != null) {
            navigationCurrentCountry.findViewById(R.id.selected).setOnClickListener(view ->
                goToView(PublicationList.class)
            );
        }

        loadList();
    }

    void loadList() {
        ((WhatTheDuckApplication) getApplication()).trackActivity(this);
        if (!hasList()) {
            this.requiresDataDownload = true;
            downloadList();
        }
        else {
            setData();
        }
    }

    private void goToView(Class<?> cls) {
        if (!ItemList.this.getClass().equals(cls)) {
            startActivity(new Intent(WhatTheDuck.wtd, cls));
        }
    }

    void goToAlternativeView() {
        type = (type.equals(WhatTheDuck.CollectionType.USER.toString()) ? WhatTheDuck.CollectionType.COA.toString() : WhatTheDuck.CollectionType.USER.toString());
        loadList();
        show();
    }

    void show() {
        if (!shouldShow()) {
            return;
        }

        Toolbar toolbar = findViewById(R.id.toolbar);
        if (shouldShowToolbar()) {
            toolbar.setVisibility(View.VISIBLE);
            setSupportActionBar(toolbar);
        }
        else {
            toolbar.setVisibility(GONE);
        }

        showNavigation();

        FloatingActionMenu addToCollection = this.findViewById(R.id.addToCollectionWrapper);
        if (shouldShowAddToCollectionButton()) {
            if (addToCollection != null) {
                addToCollection.setMenuButtonColorNormalResId(R.color.holo_green_dark);
                addToCollection.setMenuButtonColorPressedResId(R.color.holo_green_dark);
                addToCollection.setVisibility(type.equals(WhatTheDuck.CollectionType.USER.toString()) ? VISIBLE : GONE);
                addToCollection.close(false);

                if (type.equals(WhatTheDuck.CollectionType.USER.toString())) {
                    FloatingActionButton addToCollectionByPhotoButton = this.findViewById(R.id.addToCollectionByPhotoButton);
                    addToCollectionByPhotoButton.setOnClickListener(view ->
                        takeCoverPicture()
                    );

                    FloatingActionButton addToCollectionBySelectionButton = this.findViewById(R.id.addToCollectionBySelectionButton);
                    addToCollectionBySelectionButton.setOnClickListener(view -> {
                        addToCollection.setVisibility(GONE);
                        ItemList.this.goToAlternativeView();
                    });

                    Settings.saveSettings();
                }
            }
        }
        else {
            addToCollection.setVisibility(GONE);
        }

        ItemAdapter<Item> itemAdapter = getItemAdapter();
        if (this.requiresDataDownload) {
            itemAdapter.resetItems();
        }

        java.util.List<Item> items = itemAdapter.getItems();
        TextView emptyListText = this.findViewById(R.id.emptyList);
        if (emptyListText != null) {
            emptyListText.setVisibility(this.requiresDataDownload || items.size() > 0 ? INVISIBLE : VISIBLE);
        }

        RecyclerView recyclerView = findViewById(R.id.itemList);
        recyclerView.setAdapter(itemAdapter);

        EditText filterEditText = this.findViewById(R.id.filter);
        if (shouldShowFilter(items)) {
            itemAdapter.addOrReplaceFilterOnChangeListener(filterEditText);
        }
        else {
            filterEditText.setVisibility(GONE);
            itemAdapter.updateFilteredList("");
        }

        while (recyclerView.getItemDecorationCount() > 0) {
            recyclerView.removeItemDecorationAt(0);
        }
        if (hasDividers()) {
            DividerItemDecoration dividerDecoration = new DividerItemDecoration(
                recyclerView.getContext(),
                new LinearLayoutManager(this).getOrientation()
            );
            recyclerView.addItemDecoration(dividerDecoration);
        }
    }

    private void takeCoverPicture() {
        Intent takePictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        if (takePictureIntent.resolveActivity(getPackageManager()) != null) {
            CoverFlowFileHandler.current = new CoverFlowFileHandler(new WeakReference<>(this));
            Uri photoURI = CoverFlowFileHandler.current.createEmptyFileForCamera(ItemList.this);
            takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, photoURI);
            startActivityForResult(takePictureIntent, REQUEST_IMAGE_CAPTURE);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_IMAGE_CAPTURE && resultCode == RESULT_OK) {
            this.findViewById(R.id.addToCollectionWrapper).setVisibility(View.VISIBLE);
            this.findViewById(R.id.progressBar).setVisibility(VISIBLE);

            CoverFlowFileHandler.current.resizeUntilFileSize(new CoverFlowFileHandler.SearchFromCover());
        }

    }

    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_user, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        Intent i = null;

        switch (item.getItemId()) {
            case R.id.action_logout:
                Settings.setUsername(null);
                Settings.setPassword(null);
                Settings.saveSettings();
                i = new Intent(WhatTheDuck.wtd, WhatTheDuck.class);

                break;
            case R.id.action_about:
                WhatTheDuck.showAbout(this);

        }
        if (item.getTitle().equals(getString(R.string.add_issue))) {
            onBackFromAddIssueActivity();
        }
        if (i == null) {
            return super.onOptionsItemSelected(item);
        }
        else {
            startActivity(i);
            return true;
        }
    }

    private void showNavigation() {
        if (shouldShowNavigationCountry()) {
            WhatTheDuck.appDB.inducksCountryDao().findByCountryCode(WhatTheDuck.getSelectedCountry())
                .observe(this, inducksCountryName ->
                    setNavigationCountry(inducksCountryName.getCountryCode(), inducksCountryName.getCountryName())
                );
        }
        else {
            this.findViewById(R.id.navigationCountry).setVisibility(INVISIBLE);
        }
        if (shouldShowNavigationPublication()) {
            WhatTheDuck.appDB.inducksPublicationDao().findByPublicationCode(WhatTheDuck.getSelectedPublication())
                .observe(this, inducksPublication ->
                    setNavigationPublication(inducksPublication.getPublicationCode(), inducksPublication.getTitle())
                );
        }
        else {
            this.findViewById(R.id.navigationPublication).setVisibility(INVISIBLE);
        }
    }

    private void setNavigationCountry(String selectedCountry, String countryFullName) {
        final View countryNavigationView = this.findViewById(R.id.navigationCountry);
        String uri = "@drawable/flags_" + selectedCountry;
        int imageResource = getResources().getIdentifier(uri, null, getPackageName());

        if (imageResource == 0) {
            imageResource = R.drawable.flags_unknown;
        }

        ImageView currentCountryFlag = countryNavigationView.findViewById(R.id.selectedBadgeImage);
        currentCountryFlag.setImageResource(imageResource);

        TextView currentCountryText = countryNavigationView.findViewById(R.id.selectedText);
        currentCountryText.setText(countryFullName);
    }

    private void setNavigationPublication(String selectedPublication, String publicationFullName) {
        final View publicationNavigationView = this.findViewById(R.id.navigationPublication);

        TextView currentPublicationBadgeText = publicationNavigationView.findViewById(R.id.selectedBadge);
        currentPublicationBadgeText.setText(selectedPublication.split("/")[1]);

        TextView currentPublicationText = publicationNavigationView.findViewById(R.id.selectedText);
        currentPublicationText.setText(publicationFullName);
    }

    protected void storeItemList(List<Item> items) {
        this.data = items;
        this.requiresDataDownload = false;
        this.show();
    }

    void onBackFromAddIssueActivity() {
        if (isPossessedByUser()) {
            goToAlternativeView();
        }
        else {
            type = WhatTheDuck.CollectionType.USER.toString();
            startActivity(new Intent(WhatTheDuck.wtd, CountryList.class));
        }
    }

    @Override
    public boolean onContextItemSelected(MenuItem item) {
        return super.onContextItemSelected(item);
    }
}

File: app/src/main/java/net/ducksmanager/whattheduck/ItemList.kt
package net.ducksmanager.whattheduck

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.provider.MediaStore
import android.view.View
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.github.clans.fab.FloatingActionButton
import com.github.clans.fab.FloatingActionMenu
import net.ducksmanager.persistence.models.coa.InducksCountryName
import net.ducksmanager.persistence.models.coa.InducksPublication
import net.ducksmanager.util.AppCompatActivityWithDrawer
import net.ducksmanager.util.CoverFlowFileHandler
import net.ducksmanager.util.CoverFlowFileHandler.SearchFromCover
import java.lang.ref.WeakReference
import java.util.*

abstract class ItemList<Item> : AppCompatActivityWithDrawer() {

    companion object {
        @JvmField
        var type = WhatTheDuck.CollectionType.USER.toString()
        const val MIN_ITEM_NUMBER_FOR_FILTER = 20
        private const val REQUEST_IMAGE_CAPTURE = 1
    }

    private var requiresDataDownload = false

    @JvmField
    var data: List<Item> = ArrayList()

    protected abstract fun hasList(): Boolean
    protected abstract fun downloadList(currentActivity: Activity)
    protected abstract fun hasDividers(): Boolean
    protected abstract val isPossessedByUser: Boolean
    protected abstract fun setData()
    protected abstract fun shouldShow(): Boolean
    protected abstract fun shouldShowNavigationCountry(): Boolean
    protected abstract fun shouldShowNavigationPublication(): Boolean
    protected abstract fun shouldShowAddToCollectionButton(): Boolean
    protected abstract fun shouldShowFilter(items: List<Item>): Boolean

    protected abstract val itemAdapter: ItemAdapter<Item>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.wtd_list)

        showToolbarIfExists()

        findViewById<View>(R.id.navigationAllCountries)
            ?.setOnClickListener { _: View? -> goToView(CountryList::class.java) }

        findViewById<View>(R.id.navigationCountry)
            ?.findViewById<View>(R.id.selected)?.setOnClickListener { _: View? -> goToView(PublicationList::class.java) }

        loadList()
    }

    fun loadList() {
        (application as WhatTheDuck).trackActivity(this)
        if (!hasList()) {
            requiresDataDownload = true
            downloadList(this)
        } else {
            setData()
        }
    }

    private fun goToView(cls: Class<*>) {
        if (this@ItemList.javaClass != cls) {
            startActivity(Intent(this, cls))
        }
    }

    private fun goToAlternativeView() {
        type = if (type == WhatTheDuck.CollectionType.USER.toString())
            WhatTheDuck.CollectionType.COA.toString()
        else
            WhatTheDuck.CollectionType.USER.toString()
        loadList()
        show()
    }

    fun show() {
        if (!shouldShow()) {
            return
        }
        showNavigation()
        val addToCollection = findViewById<FloatingActionMenu>(R.id.addToCollectionWrapper)
        if (shouldShowAddToCollectionButton()) {
            if (addToCollection != null) {
                addToCollection.setMenuButtonColorNormalResId(R.color.holo_green_dark)
                addToCollection.setMenuButtonColorPressedResId(R.color.holo_green_dark)
                addToCollection.visibility = if (type == WhatTheDuck.CollectionType.USER.toString()) View.VISIBLE else View.GONE
                addToCollection.close(false)

                if (type == WhatTheDuck.CollectionType.USER.toString()) {
                    findViewById<FloatingActionButton>(R.id.addToCollectionByPhotoButton)
                        .setOnClickListener { takeCoverPicture() }

                    findViewById<FloatingActionButton>(R.id.addToCollectionBySelectionButton)
                        .setOnClickListener {
                            addToCollection.visibility = View.GONE
                            goToAlternativeView()
                        }
                }
            }
        } else {
            addToCollection!!.visibility = View.GONE
        }
        val itemAdapter = itemAdapter

        if (requiresDataDownload) {
            itemAdapter.resetItems()
        }

        val recyclerView = findViewById<RecyclerView>(R.id.itemList)
        recyclerView.adapter = itemAdapter

        val filterEditText = findViewById<EditText>(R.id.filter)
        if (shouldShowFilter(itemAdapter.items)) {
            itemAdapter.addOrReplaceFilterOnChangeListener(filterEditText)
        } else {
            filterEditText.visibility = View.GONE
            itemAdapter.updateFilteredList("")
        }
        findViewById<TextView>(R.id.emptyList)
            ?.visibility = if (requiresDataDownload || itemAdapter.itemCount > 0) View.INVISIBLE else View.VISIBLE

        while (recyclerView.itemDecorationCount > 0) {
            recyclerView.removeItemDecorationAt(0)
        }
        if (hasDividers()) {
            recyclerView.addItemDecoration(DividerItemDecoration(
                recyclerView.context,
                LinearLayoutManager(this).orientation
            ))
        }
    }

    private fun takeCoverPicture() {
        val takePictureIntent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        if (takePictureIntent.resolveActivity(packageManager) != null) {
            CoverFlowFileHandler.current = CoverFlowFileHandler(WeakReference(this))

            val photoURI = CoverFlowFileHandler.current.createEmptyFileForCamera(this@ItemList)
            takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, photoURI)

            startActivityForResult(takePictureIntent, REQUEST_IMAGE_CAPTURE)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_IMAGE_CAPTURE && resultCode == Activity.RESULT_OK) {
            findViewById<View>(R.id.addToCollectionWrapper).visibility = View.VISIBLE
            findViewById<View>(R.id.progressBar).visibility = View.VISIBLE
            CoverFlowFileHandler.current.resizeUntilFileSize(SearchFromCover())
        }
    }

    private fun showNavigation() {
        if (shouldShowNavigationCountry()) {
            WhatTheDuck.appDB.inducksCountryDao().findByCountryCode(WhatTheDuck.selectedCountry)
                .observe(this, Observer { inducksCountryName: InducksCountryName ->
                    setNavigationCountry(inducksCountryName.countryCode, inducksCountryName.countryName)
                })
        } else {
            findViewById<View>(R.id.navigationCountry).visibility = View.INVISIBLE
        }
        if (shouldShowNavigationPublication()) {
            WhatTheDuck.appDB.inducksPublicationDao().findByPublicationCode(WhatTheDuck.selectedPublication!!)
                .observe(this, Observer { inducksPublication: InducksPublication ->
                    setNavigationPublication(inducksPublication.publicationCode, inducksPublication.title)
                })
        } else {
            findViewById<View>(R.id.navigationPublication).visibility = View.INVISIBLE
        }
    }

    private fun setNavigationCountry(selectedCountry: String, countryFullName: String) {
        val countryNavigationView = findViewById<View>(R.id.navigationCountry)
        val uri = "@drawable/flags_$selectedCountry"
        var imageResource = resources.getIdentifier(uri, null, packageName)
        if (imageResource == 0) {
            imageResource = R.drawable.flags_unknown
        }

        countryNavigationView.findViewById<ImageView>(R.id.selectedBadgeImage)
            .setImageResource(imageResource)
        countryNavigationView.findViewById<TextView>(R.id.selectedText)
            .text = countryFullName
    }

    private fun setNavigationPublication(selectedPublication: String, publicationFullName: String) {
        val publicationNavigationView = findViewById<View>(R.id.navigationPublication)

        publicationNavigationView.findViewById<TextView>(R.id.selectedBadge)
            .text = selectedPublication.split("/").toTypedArray()[1]

        publicationNavigationView.findViewById<TextView>(R.id.selectedText)
            .text = publicationFullName
    }

    fun storeItemList(items: List<Item>) {
        data = items
        requiresDataDownload = false
        show()
    }

    fun onBackFromAddIssueActivity() {
        if (isPossessedByUser) {
            goToAlternativeView()
        } else {
            type = WhatTheDuck.CollectionType.USER.toString()
            startActivity(Intent(this, CountryList::class.java))
        }
    }
}